File: benchmarks/tsb/bench_period_asfreq.ts
/**
 * Benchmark: Period.asfreq and PeriodIndex.asfreq — frequency conversion.
 * Outputs JSON: {"function": "period_asfreq", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Period, PeriodIndex } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 3;
const ITERATIONS = 20;

// Build a PeriodIndex of monthly periods using periodRange
const startMonth = Period.fromString("2000-01", "M");
const idx = PeriodIndex.periodRange(startMonth, SIZE);

for (let i = 0; i < WARMUP; i++) {
  idx.asfreq("D", "start");
  idx.asfreq("D", "end");
  idx.asfreq("Q", "start");
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  idx.asfreq("D", "start");
  idx.asfreq("D", "end");
  idx.asfreq("Q", "start");
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "period_asfreq",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_period_asfreq.py
"""
Benchmark: pandas Period.asfreq and PeriodIndex.asfreq — frequency conversion.
Outputs JSON: {"function": "period_asfreq", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

SIZE = 10_000
WARMUP = 3
ITERATIONS = 20

idx = pd.period_range(start="2000-01", periods=SIZE, freq="M")

for _ in range(WARMUP):
    idx.asfreq("D", how="start")
    idx.asfreq("D", how="end")
    idx.asfreq("Q", how="start")

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    idx.asfreq("D", how="start")
    idx.asfreq("D", how="end")
    idx.asfreq("Q", how="start")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "period_asfreq", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))
